signer/core/signed_data.go

```diff
@@ -0,0 +1,633 @@
+// signing of transactions.
+// hash = keccak256("\x19$Ethereum Signed Message:\n"${message length}${message}).
+func SignTextPlain(data hexutil.Bytes) (hexutil.Bytes, string) {
+	// The letter `E` is \x45 in hex, retrofitting
+	// https://github.com/ethereum/go-ethereum/pull/2940/commits
+	msg := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(data), string(data))
+	return crypto.Keccak256([]byte(msg)), msg
+}
+
+// SignTypedData signs EIP-712 conformant typed data
+// hash = keccak256("\x19${byteVersion}${domainSeparator}${hashStruct(message)}")
+func (api *SignerAPI) SignTypedData(ctx context.Context, addr common.MixedcaseAddress, typedData TypedData) (hexutil.Bytes, error) {
+	domainSeparator := typedData.HashStruct("EIP712Domain", typedData.Domain.Map())
+	typedDataHash := typedData.HashStruct(typedData.PrimaryType, typedData.Message)
+	_, err := json.Marshal(typedData.Map())
+	if err != nil {
+		return nil, err
+	}
+	msg := fmt.Sprintf("\x19\x01%s%s", string(domainSeparator), string(typedDataHash))
+	sighash := crypto.Keccak256([]byte(msg))
+	req := &SignDataRequest{Rawdata: typedData.Map(), Message: msg, Hash: sighash, ContentType: DataTyped.Mime}
+	signature, err := api.Sign(ctx, addr, req)
+	if err != nil {
+		api.UI.ShowError(err.Error())
+		return nil, err
+	}
+	return signature, nil
+}
+
+// HashStruct generates the following encoding for the given domain and message:
+// `encode(domainSeparator : 𝔹²⁵⁶, message : 𝕊) = "\x19\x01" ‖ domainSeparator ‖ hashStruct(message)`
+func (typedData *TypedData) HashStruct(primaryType string, data EIP712Data) hexutil.Bytes {
+	return crypto.Keccak256(typedData.EncodeData(primaryType, data))
+}
+
+// Dependencies returns an array of custom types ordered by their hierarchical reference tree
+func (typedData *TypedData) Dependencies(primaryType string, found []string) []string {
+	includes := func(arr []string, str string) bool {
+		for _, obj := range arr {
+			if obj == str {
+				return true
+			}
+		}
+		return false
+	}
+
+	if includes(found, primaryType) {
+		return found
+	}
+	if typedData.Types[primaryType] == nil {
+		return found
+	}
+	found = append(found, primaryType)
+	for _, field := range typedData.Types[primaryType] {
+		for _, dep := range typedData.Dependencies(field["type"], found) {
+			if !includes(found, dep) {
+				found = append(found, dep)
+			}
+		}
+	}
+	return found
+}
+
+// EncodeType generates the following encoding:
+// `name ‖ "(" ‖ member₁ ‖ "," ‖ member₂ ‖ "," ‖ … ‖ memberₙ ")"`
+//
+// each member is written as `type ‖ " " ‖ name` encodings cascade down and are sorted by name
+func (typedData *TypedData) EncodeType(primaryType string) hexutil.Bytes {
+	// Get dependencies primary first, then alphabetical
+	deps := typedData.Dependencies(primaryType, []string{})
+	for i, dep := range deps {
+		if dep == primaryType {
+			deps = append(deps[:i], deps[i+1:]...)
+			break
+		}
+	}
+	deps = append([]string{primaryType}, deps...)
```

Comment:
@Recmo interesting! Could you give an example of such as struct, so we can add it as a testcase ?

Reply:
```solidity
struct LinkedList {
   bytes32 contents;
   LinkedList next;
}
```

Actually, it's a a moot point, as Solidity itself doesn't support recursive data types.

Reply:
If there is such a thing, it would be good to have a testcase for it, to ensure that we don't `panic` or `OOM` or something like that.

Reply:
Indeed it's not handled, but it could be by making the array search exhaustive by simply removing the `break` in the for loop. However, as per @Recmo's latest comment, I'll just leave that as it is for now but write some tests to make sure.

Reply:
Sorry, my bad. It is actually handled correctly due to how Dependencies is implemented. It will never add the same item twice.

In practice, due to how `Dependencies` is implemented and called, the `primaryType` will always be the first element, so the loop in `EncodeType` will always terminate on the first item.

In other words, you could also do this:

```
deps := typedData.Dependencies(primaryType, []string{})
deps = append([]string{primaryType}, deps[1:])
```

Reply:
It's missing the sorting step. All the dependencies following the primary one should be sorted alphabetically. This would make it something like:

```
deps := typedData.Dependencies(primaryType, []string{})
deps = append([]string{primaryType}, sortAlphabetically(deps[1:]))
```

This should be fixed or signatures will fail.

Reply:
Oops, fixed in the last commit!

Reply:
@Recmo I just remembered that recursive types are automatically rejected by the latest implementation due to this snippet (lines 620-622):

```
if typeKey == typeVal {
	return fmt.Errorf("type '%s' cannot reference itself", typeVal)
}
```

Would you like this to be removed?